### PR TITLE
Testing version Qv2ray v2.1

### PR DESCRIPTION
**Pull request by Qv2ray-dev**
Qv2ray v2.1

### CODE REVIEW

**Comment by Qv2ray-dev** [comment]
As there’re so many changes on this merge, the version number will be bumped to Qv2ray v2.1

**Comment by DuckSoft** [comment]
> As there’re so many changes on this merge, the version number will be bumped to Qv2ray v2.1

I agree.

**Comment by DuckSoft** [comment]
Let's merge and translate, will we? We will be able to release Qv2ray v2.1 before daybreak.

**Comment by Qv2ray-dev** [comment]
Not ready yet, I don’t want to leave known bugs this time

**Comment by DuckSoft** [comment]
> Not ready yet, I don’t want to leave known bugs this time

Setting #302 as a blocker?

**Comment by Qv2ray-dev** [comment]
Yes, sure

**Comment by DuckSoft** [comment]
> Yes, sure

Set

File: src/base/Qv2rayLog.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#include "Qv2rayLog.hpp"
+#include "GlobalInstances.hpp"
+
+namespace Qv2ray::base
+{
+    // Forwarded from QvTinyLog
+    static QQueue<QString> __loggerBuffer;
```
Comment (DuckSoft) [inline src/base/Qv2rayLog.cpp]: Seems that `QQueue<>` is not thread-safe.
Maybe you need something like `QMutex` & `QMutexLocker` to make it safe.

File: src/common/CommandArgs.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        //
+        parser.addOption(noAPIOption);
+        parser.addOption(runAsRootOption);
+        parser.addOption(debugOption);
+        parser.addOption(withToolbarOption);
+        helpOption = parser.addHelpOption();
+        versionOption = parser.addVersionOption();
+    }
+
+    CommandLineParseResult QvCommandArgParser::ParseCommandLine(QString *errorMessage)
```
Comment (DuckSoft) [inline src/common/CommandArgs.cpp]: try make good use of `std::optional<QString>` as return type in C++17.

File: src/components/tcping/QvTCPing.hpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Qv2ray::components::tcping
+{
+    struct QvTCPingData {
+        ConnectionIdentifier connectionIdentifier;
+        QString hostName;
+        int port;
+        QString errorMessage;
+        int total, succeed, failed;
+        double min = 999999999999999.0, max = 0.0, avg = 0.0;
```
Comment (DuckSoft) [inline src/components/tcping/QvTCPing.hpp]: import `<limits>` and use `std::numeric_limits<double>::max()` to replace this ugly thing.

File: src/core/kernel/KernelInteractions.cpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
+#include <QObject>
+#include <QWidget>
+#include <QDesktopServices>
+#include "common/QvHelpers.hpp"
+#include "KernelInteractions.hpp"
+#include "core/connection/ConnectionIO.hpp"
+
+namespace Qv2ray::core::kernel
+{
+    bool V2rayKernelInstance::ValidateKernel(const QString &vCorePath, const QString &vAssetsPath, QString *message)
```
Comment (DuckSoft) [inline src/core/kernel/KernelInteractions.cpp]: again: try replace with `std::optional<QString>` as return type

File: src/common/CommandArgs.cpp (hunk 2 above)
Comment (DuckSoft) [reply src/common/CommandArgs.cpp]: example:
```cpp
std::optional<QString> check() {
  if (something_worked()) return {};
  else return "something is wrong";
}
```

Then use it like this:
```cpp
// when returned nothing, if fails automatically
if (auto errMsg = check()) {
  std::cout << "crap! some error happened: " << errMsg << std::endl;
} else {
  std::cout << "everything is just fine" << std::endl;
}
```

File: src/components/tcping/QvTCPing.hpp (hunk 3 above)
Comment (Qv2ray-dev) [reply src/components/tcping/QvTCPing.hpp]: ack

Comment (Qv2ray-dev) [reply src/components/tcping/QvTCPing.hpp]: It was initially ported from C, which was what the original author did.
We’ll make good use of c++17

File: src/core/kernel/KernelInteractions.cpp (hunk 4 above)
Comment (Qv2ray-dev) [reply src/core/kernel/KernelInteractions.cpp]: I think we need the bool as the return type to check if the check is passed.
Or use &bool parameter and optional string as the return type?

File: src/common/CommandArgs.cpp (hunk 2 above)
Comment (Qv2ray-dev) [reply src/common/CommandArgs.cpp]: This was directly copied from the Qt examples, we need the enums to determine the exit code if the command arg parse failed.

File: src/base/models/QvRuntimeConfig.hpp [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#define SET_RUNTIME_CONFIG(conf, val) RuntimeConfig.conf = val();
+#define RESTORE_RUNTIME_CONFIG(conf, func) func(RuntimeConfig.conf);
+
+namespace Qv2ray::base
+{
+    struct Qv2rayRuntimeConfig {
+        /// Whether to hide Qv2ray when taking a screenshot
+        bool screenShotHideQv2ray;
```
Comment (DuckSoft) [inline src/base/models/QvRuntimeConfig.hpp]: initial value maybe undefined behaviour. try specify initial value explicitly.

File: src/components/tcping/QvTCPing.hpp (hunk 3 above)
Comment (Qv2ray-dev) [reply src/components/tcping/QvTCPing.hpp]: This is already fixed.

File: src/base/models/QvRuntimeConfig.hpp (hunk 5 above)
Comment (Qv2ray-dev) [reply src/base/models/QvRuntimeConfig.hpp]: fixed

**Comment by DuckSoft** [comment]
Awaiting Upstream Package for release:
https://github.com/microsoft/vcpkg/pull/9963

**Comment by Qv2ray-dev** [comment]
All blockers are solved so this can be reviewed and merged.

**Comment by Qv2ray-dev** [comment]
Merging.

**Pull request merged**